common/src/lib.rs

```diff
@@ -182,12 +207,31 @@ pub fn wait_key(prog_id: H256, msg_id: H256) -> Vec<u8> {
     key
 }
 
-pub fn get_code(code_hash: H256) -> Option<Vec<u8>> {
-    sp_io::storage::get(&code_key(code_hash).0)
+// TODO [SAB] refactor to write all the struct somehow? Maybe separate 1) Metadata from 2) code and write each...
+
+pub fn get_code(code_hash: H256) -> Option<CodeWithMetadata> {
+    sp_io::storage::get(&code_key(code_hash).0).map(|data| {
+        let author = H256::from_slice(&data[..32]);
+        let block_number = {
+            let mut bytes = [0u8; 4];
+            bytes.copy_from_slice(&data[32..36]);
+            u32::from_le_bytes(bytes)
+        };
+        let code = (&data[36..]).to_vec();
+        CodeWithMetadata {
+            code,
+            author,
+            block_number,
+        }
+    })
 }
 
-pub fn set_code(code_hash: H256, code: &[u8]) {
-    sp_io::storage::set(&code_key(code_hash).0, code)
+pub fn set_code(code_hash: H256, code: CodeWithMetadata) {
+    let mut data = Vec::new();
+    code.author.encode_to(&mut data);
```

Comment:
Doesn't this mean that if the same piece of code is submitted multiple times the metadata will be overwritten every time and only contain the details of who (and when) has submitted it last? This might render the TTL-based logic potentially susceptible to misuse (although, at obvious expense for an attacker).

Reply:
@ekovalev well it depends on how `set_code` is actually used. Calls to the function are performed from `set_program` and `submit_code`. In the first one the code is checked for existence, i.e. If it exists we don't save it. In the second function (extrinsic) code can be set only if it doesn't exist. So, no double submissions - no rewrites of metadata.

Should be stated, that this metadata is only needed to invalidate "old" code data entries either by using authors address (charge him) or block number (TTL).

Reply:
Oh, btw, what do you think about putting metadata to other prefix in storage?

Reply:
> Oh, btw, what do you think about putting metadata to other prefix in storage?

Please see my comment above. In short, I'd definitely do so.